scipy/linalg/_decomp_cossin.py

```diff
@@ -0,0 +1,186 @@
+from collections import Iterable
+
+import numpy as np
+from numpy.linalg import LinAlgError
+
+from . import block_diag
+from .lapack import _compute_lwork, get_lapack_funcs
+
+__all__ = ['cossin']
+
+from .._lib._util import _asarray_validated
+
+
+def cossin(X, partitioning=None):
```

Comment:
I am a bit undecided on this. Let's take two scenarios; 
 - I have the subarrays and I want to get the decomposition. What I would really like is to slap a dict or list with `cossin([a,b,c,d])` and be done. 
  - I have big matrix and I'd like to just type `cossin(X, 3, 5)` and be done with it without the trouble of iterables. Also for us we just pass the input through `int()` if they are not None so no funky iterable checks.

Thus, I would argue that `cossin(X, p=None, q=None, compute_u=True, compute_vh=True)` signature seems more usable to me than entering this way but it's just me so I'm open for changing my mind.

Reply:
I am going to trust your instincts on this if you say this would be more numpy/scipy-esque. 

The one small reason I'd mention why created the single tuple for partitioning is because in the end we are specifying a shape (of the upper-left block), which is not uncommon in numpy/scipy world as an arg (though admittedly annoying with the parentheses), also p without q or vice versa doesn't make sense, they just make sense together - unless maybe `cossin(X, 3)` would mean `cossin(X, 3, 1)`? What do you think?

Regarding `compute_u`, `compute_vh` - I guess those could be interesting to save time/performance, so I'll add them in.